README.md
# Commodity [![crates.io badge](https://img.shields.io/crates/v/commodity.svg)](https://crates.io/crates/commodity) [![docs.rs badge](https://docs.rs/commodity/badge.svg)](https://docs.rs/commodity/) [![license badge](https://img.shields.io/github/license/kellpossible/commodity)](https://github.com/kellpossible/commodity/blob/master/LICENSE.txt) [![github action badge](https://github.com/kellpossible/commodity/workflows/Rust/badge.svg)](https://github.com/kellpossible/commodity/actions?query=workflow%3ARust)

A library for representing commodities/currencies, and exchange rates/conversions between them in Rust. Values are backed by the [rust_decimal](https://crates.io/crates/rust_decimal) library.

**[Changelog](./CHANGELOG.md)**

## Optional Features

The following features can be enabled to provide extra functionality:

+ `serde-support`
  + Enables support for serialization/de-serialization via `serde`

## Example

```rust
use commodity::{Commodity, CommodityType, CommodityTypeID};
use rust_decimal::Decimal;
use std::str::FromStr;

// Create a commodity type from a currency's iso4317 three character code.
// The CommodityType stores information associated with that currency,
// such as the full name ("United States dollar" for this one).
let usd = CommodityType::from_currency_alpha3("USD").unwrap();

// Create a commodity with a value of "2.02 USD"
let commodity1 = Commodity::new(Decimal::from_str("2.02").unwrap(), &usd);

// Create commodities using the `from_str` method
let commodity2 = Commodity::from_str("24.00 USD").unwrap();

// Create commodity using a CommodityTypeID
let nzd_code = CommodityTypeID::from_str("NZD").unwrap();
let commodity3 = Commodity::new(Decimal::from_str("24.00").unwrap(), nzd_code);

// Add two compatible (same currency) commodities, the result has
// the same currency (in this case, "USD").
let commodity4 = commodity1.add(&commodity2).unwrap();

// Try to subtract two incompatible commodities
let result = commodity3.sub(&commodity2);
assert!(result.is_err());
```


CHANGELOG.md
# Changelog for Commodity

## v0.4.0

New Features:

+ Add `rate_between()` method to `ExchangeRate`.

Breaking:

+ Make `iso4217` an optional dependency (disabled by default).

## v0.3.7

+ Implement `Hash` for `CommodityType`.

## v0.3.6

+ Implement `Eq` for `CommodityType`.

## v0.3.5

+ Implement `PartialEq` for `ExchangeRate`.

## v0.3.4

+ Implement `Display` for `CommodityType`.

## v0.3.3

+ Implement `PartialEq` for `CommodityType`.

## v0.3.2

+ Fix this changelog!

## v0.3.1

+ Bump `rust_decimal` dependency up to using generic version `1` to address [#3](https://github.com/kellpossible/commodity/issues/3).
+ CommodityType#from_str() now accepts &str or String for arguments using AsRef\<str\> and Into\<String\> to address [#2](https://github.com/kellpossible/commodity/issues/2).
+ CommodityType#from_currency_alpha3() now accepts &str or String using AsRef\<str\> to address [#2](https://github.com/kellpossible/commodity/issues/2).

## v0.3.0

+ Renamed `Currency` to `CommodityType` and associated member variables to address issue #1.
+ Renamed `CurrencyCode` to `CommodityTypeID` and associated member variables.
+ Renamed errors associated with `CommodityType` and `CommodityTypeID`.


src/exchange_rate/mod.rs
//! Types and utilities relating to exchange rates and conversions
//! between different types of commodities.

use crate::{Commodity, CommodityTypeID};
use chrono::{DateTime, NaiveDate, Utc};
use rust_decimal::Decimal;

#[cfg(feature = "serde-support")]
use serde::{Deserialize, Serialize};

use std::collections::BTreeMap;
use thiserror::Error;

/// An error associated with functionality in the [exchange_rate](crate::exchange_rate) module.
#[derive(Error, Debug)]
pub enum ExchangeRateError {
    #[error("The commodity type with id {0} is not present in the exchange rate.")]
    CommodityTypeNotPresent(CommodityTypeID),
    #[error("There was a divide overflow while computing the exchange rate, performing the division {0}/{1}.")]
    DivideOverflow(Decimal, Decimal),
}

/// Represents the exchange rate between [Commodity](Commodity)s
/// with different [CommodityType](crate::CommodityType)s.
#[cfg_attr(feature = "serde-support", derive(Serialize, Deserialize))]
#[derive(Debug, Clone, PartialEq)]
pub struct ExchangeRate {
    /// The datetime that this exchange rate represents
    pub date: Option<NaiveDate>,
    /// The datetime that this exchange rate was obtained.
    pub obtained_datetime: Option<DateTime<Utc>>,
    /// The id of the base commodity type for the exchange rate
    pub base: Option<CommodityTypeID>,
    /// Maps commodity type ids, to the conversion rate from that
    /// [CommodityType](crate::CommodityType) to the `base`
    /// [CommodityType](crate::CommodityType).
    pub rates: BTreeMap<CommodityTypeID, Decimal>,
}

impl ExchangeRate {
    pub fn get_rate(&self, commodity_type_id: &CommodityTypeID) -> Option<&Decimal> {
        self.rates.get(commodity_type_id)
    }

    /// Convert the [CommodityType](crate::CommodityType) of a
    /// [Commodity](Commodity) to another
    /// [CommodityType](crate::CommodityType) using this
    /// [ExchangeRate](ExchangeRate).
    pub fn convert(
        &self,
        commodity: Commodity,
        target_commodity_type: CommodityTypeID,
    ) -> Result<Commodity, ExchangeRateError> {
        if let Some(base) = self.base {
            if commodity.type_id == base {
                if let Some(rate) = self.get_rate(&target_commodity_type) {
                    return Ok(Commodity::new(
                        rate * commodity.value,
                        target_commodity_type,
                    ));
                };
            }

            if target_commodity_type == base {
                if let Some(rate) = self.get_rate(&commodity.type_id) {
                    let div = commodity
                        .value
                        .checked_div(*rate)
                        .ok_or_else(|| ExchangeRateError::DivideOverflow(commodity.value, *rate))?;
                    return Ok(Commodity::new(div, target_commodity_type));
                };
            }
        }

        // handle the situation where there is no base commodity type, or neither the commodity
        // type or the target commodity type are the base commodity type.

        let commodity_rate = match self.get_rate(&commodity.type_id) {
            Some(rate) => rate,
            None => {
                return Err(ExchangeRateError::CommodityTypeNotPresent(
                    commodity.type_id,
                ))
            }
        };

        let target_rate = match self.get_rate(&target_commodity_type) {
            Some(rate) => rate,
            None => {
                return Err(ExchangeRateError::CommodityTypeNotPresent(
                    target_commodity_type,
                ))
            }
        };

        let div = commodity
            .value
            .checked_div(*commodity_rate)
            .ok_or_else(|| ExchangeRateError::DivideOverflow(commodity.value, *commodity_rate))?;
        let value = div * target_rate;

        Ok(Commodity::new(value, target_commodity_type))
    }

    /// Get the exchange rate between two commodity types present in this exchange
    /// rate data structure. Returns `None` if one of the commodity types is not present.
    pub fn rate_between(
        &self,
        from: &CommodityTypeID,
        to: &CommodityTypeID,
    ) -> Result<Option<Decimal>, ExchangeRateError> {
        if let Some(base) = &self.base {
            if from == base {
                if let Some(rate) = self.get_rate(&to) {
                    return Ok(Some(*rate));
                };
            }

            if to == base {
                if let Some(rate) = self.get_rate(&from) {
                    let one = Decimal::new(1, 0);
                    return match one.checked_div(*rate) {
                        Some(value) => Ok(Some(value)),
                        None => Err(ExchangeRateError::DivideOverflow(one, *rate)),
                    };
                };
            }
        }

        // handle the situation where there is no base commodity type, or neither the from
        // or the to commodity types are the base commodity type.

        let from_rate = match self.get_rate(&from) {
            Some(rate) => rate,
            None => return Ok(None),
        };

        let to_rate = match self.get_rate(&to) {
            Some(rate) => rate,
            None => return Ok(None),
        };

        match to_rate.checked_div(*from_rate) {
            Some(value) => Ok(Some(value)),
            None => Err(ExchangeRateError::DivideOverflow(*to_rate, *from_rate)),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{Commodity, CommodityTypeID, ExchangeRate};
    use chrono::NaiveDate;
    use rust_decimal::Decimal;
    use std::collections::BTreeMap;
    use std::str::FromStr;

    #[cfg(feature = "serde-support")]
    #[test]
    fn test_json_serialization() {
        use serde_json;

        let original_data = r#"{
    "date": "2020-02-07",
    "base": "AUD",
    "rates": {
        "USD": 2.542,
        "EU": "1.234"
    }
}
"#;

        let exchange_rate: ExchangeRate = serde_json::from_str(original_data).unwrap();
        let usd = CommodityTypeID::from_str("USD").unwrap();
        let eu = CommodityTypeID::from_str("EU").unwrap();

        assert_eq!(
            NaiveDate::from_ymd(2020, 02, 07),
            exchange_rate.date.unwrap()
        );
        assert_eq!("AUD", exchange_rate.base.unwrap());
        assert_eq!(
            Decimal::from_str("2.542").unwrap(),
            *exchange_rate.get_rate(&usd).unwrap()
        );
        assert_eq!(
            Decimal::from_str("1.234").unwrap(),
            *exchange_rate.get_rate(&eu).unwrap()
        );

        let expected_serialized_data = r#"{
  "date": "2020-02-07",
  "obtained_datetime": null,
  "base": "AUD",
  "rates": {
    "EU": "1.234",
    "USD": "2.542"
  }
}"#;

        let serialized_data = serde_json::to_string_pretty(&exchange_rate).unwrap();
        assert_eq!(expected_serialized_data, serialized_data);
    }

    /// Convert between two commodities at a reference rate (no base rate commodity type).
    #[test]
    fn convert_reference_rates() {
        let mut rates: BTreeMap<CommodityTypeID, Decimal> = BTreeMap::new();
        let aud = CommodityTypeID::from_str("AUD").unwrap();
        let nzd = CommodityTypeID::from_str("NZD").unwrap();
        rates.insert(aud, Decimal::from_str("1.6417").unwrap());
        rates.insert(nzd, Decimal::from_str("1.7094").unwrap());

        let exchange_rate = ExchangeRate {
            date: Some(NaiveDate::from_ymd(2020, 02, 07)),
            base: None,
            obtained_datetime: None,
            rates,
        };

        {
            let start_commodity = Commodity::new(Decimal::from_str("10.0").unwrap(), aud);
            let converted_commodity = exchange_rate.convert(start_commodity, nzd);
            assert_eq!(
                Decimal::from_str("10.412377413656575501005055735").unwrap(),
                converted_commodity.unwrap().value
            );
            assert_eq!(
                exchange_rate.rate_between(&aud, &nzd).unwrap(),
                Some(Decimal::from_str("1.0412377413656575501005055734").unwrap())
            );
        }

        {
            let start_commodity = Commodity::new(Decimal::from_str("10.0").unwrap(), nzd);
            let converted_commodity = exchange_rate.convert(start_commodity, aud);
            assert_eq!(
                Decimal::from_str("9.603954603954603954603954604").unwrap(),
                converted_commodity.unwrap().value
            );
            assert_eq!(
                exchange_rate.rate_between(&nzd, &aud).unwrap(),
                Some(Decimal::from_str("0.9603954603954603954603954603").unwrap())
            );
        }
    }

    /// Convert between commodities using an exchange rate with a base rate.
    #[test]
    fn convert_base_rate() {
        let mut rates: BTreeMap<CommodityTypeID, Decimal> = BTreeMap::new();
        let nok = CommodityTypeID::from_str("NOK").unwrap();
        let usd = CommodityTypeID::from_str("USD").unwrap();
        let gel = CommodityTypeID::from_str("GEL").unwrap();

        rates.insert(nok, Decimal::from_str("9.2691220713").unwrap());
        rates.insert(gel, Decimal::from_str("3.08").unwrap());

        let exchange_rate = ExchangeRate {
            date: Some(NaiveDate::from_ymd(2020, 02, 07)),
            base: Some(usd),
            obtained_datetime: None,
            rates,
        };

        {
            let start_commodity = Commodity::new(Decimal::from_str("100.0").unwrap(), usd);
            let converted_commodity = exchange_rate.convert(start_commodity, nok);
            assert_eq!(
                Decimal::from_str("926.91220713000").unwrap(),
                converted_commodity.unwrap().value
            );
            assert_eq!(
                exchange_rate.rate_between(&usd, &nok).unwrap(),
                Some(Decimal::from_str("9.2691220713").unwrap())
            );
        }

        {
            let start_commodity = Commodity::new(Decimal::from_str("100.0").unwrap(), nok);
            let converted_commodity = exchange_rate.convert(start_commodity, usd);
            assert_eq!(
                Decimal::from_str("10.788508256853169187585300627").unwrap(),
                converted_commodity.unwrap().value
            );
            assert_eq!(
                exchange_rate.rate_between(&nok, &usd).unwrap(),
                Some(Decimal::from_str("0.1078850825685316918758530063").unwrap())
            );
        }

        {
            let start_commodity = Commodity::new(Decimal::from_str("100.0").unwrap(), nok);
            let converted_commodity = exchange_rate.convert(start_commodity, gel);
            assert_eq!(
                Decimal::from_str("33.228605431107761097762725931").unwrap(),
                converted_commodity.unwrap().value
            );
            assert_eq!(
                exchange_rate.rate_between(&nok, &gel).unwrap(),
                Some(Decimal::from_str("0.3322860543110776109776272593").unwrap())
            );
        }
    }
}


src/lib.rs
//! A library with primatives representing commodities/money.
//!
//! # Optional Features
//!
//! The commodity package has the following optional cargo features:
//!
//! + `serde-support`
//!   + Disabled by default
//!   + Enables support for serialization/de-serialization via `serde`
//!
//! # Usage
//!
//! This library revolves around the [Commodity](Commodity) struct,
//! which stores a value using
//! [rust_decimal::Decimal](rust_decimal::Decimal), and a
//! [CommodityTypeID](CommodityTypeID) which denotes the type of commodity.
//! Commodities with different currencies cannot interact with
//! mathematical operations such as `add`, `sub`, etc, this is checked
//! at runtime.
//!
//! [CommodityTypeID](CommodityTypeID) designed to be used directly when
//! when working with commodities, it is backed by a small fixed size
//! array which supports the [Copy](std::marker::Copy) trait,
//! hopefully making it easier fast, lock-free concurrent code that
//! deals with commodities.
//!
//! [CommodityType](CommodityType) is designed to store useful user-facing
//! information about the currency being referenced by the
//! [CommodityTypeID](CommodityTypeID), such as its full name/description.
//!
//! ```
//! use commodity::{Commodity, CommodityType, CommodityTypeID};
//! use rust_decimal::Decimal;
//! use std::str::FromStr;
//!
//! // Create a currency from its iso4317 three character code.
//! // The currency stores information associated with the currency,
//! // such as the full name ("United States dollar" for this one).
//! let usd = CommodityType::from_currency_alpha3("USD").unwrap();
//!
//! // Create a commodity with a value of "2.02 USD"
//! let commodity1 = Commodity::new(Decimal::from_str("2.02").unwrap(), &usd);
//!
//! // Create commodities using the `from_str` method
//! let commodity2 = Commodity::from_str("24.00 USD").unwrap();
//!
//! // Create commodity using a CommodityTypeID
//! let nzd_code = CommodityTypeID::from_str("NZD").unwrap();
//! let commodity3 = Commodity::new(Decimal::from_str("24.00").unwrap(), nzd_code);
//!
//! // Add two compatible (same currency) commodities, the result has
//! // the same currency (in this case, "USD").
//! let commodity4 = commodity1.add(&commodity2).unwrap();
//!
//! // Try to subtract two incompatible commodities
//! let result = commodity3.sub(&commodity2);
//! assert!(result.is_err());
//! ```

extern crate arrayvec;
extern crate chrono;
#[cfg(feature = "iso4217")]
extern crate iso4217;
extern crate rust_decimal;

#[cfg(feature = "serde-support")]
extern crate serde;

#[cfg(test)]
#[cfg(feature = "serde-support")]
extern crate serde_json;

mod commodity;
pub mod exchange_rate;

pub use crate::commodity::*;

#[cfg(doctest)]
#[macro_use]
extern crate doc_comment;

#[cfg(doctest)]
doctest!("../README.md");
